File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/sql/CustomersNeverOrdered.scala
package io.devorbit.sparklabs.batch.sales.sql

object CustomersNeverOrdered extends App {
  SqlUtils.sql(
    """
      |SELECT c.*
      |FROM customers c
      |LEFT ANTI JOIN orders o ON c.customer_id = o.customer_id
      |""".stripMargin
  ).show(false)
}

File: python/sparklabs/batch/sales/sql/customers_never_ordered.py
from sparklabs.batch.sales.sql import utils


def run() -> None:
    utils.sql(
        """
        SELECT c.*
        FROM customers c
        LEFT ANTI JOIN orders o ON c.customer_id = o.customer_id
        """
    ).show(truncate=False)


if __name__ == "__main__":
    run()
